### ISSUE TITLE
detach command

### ISSUE THREAD

**Issue by 73** [pubs/pubs#41]
Hi,
is there a way to `detach` a document after it was attached? I think it doesn't necessarily have to be a command. Something like `pubs attach --remove` would suffice.
-- 73

**Comment by humm** [comment]
Yes, this needs to be implemented. As well as a way to copy the attached document to the current directory.

**Comment by 73** [comment]
What do you mean? Like `pubs export -d citekey`?
That should be another issue. Should it not?
Does it make sense to use a 0.6.0 milestone and assign issues to that milestone?

**Comment by 73** [comment]
Okay. I found two possibilities to be viable:
1. Implement a `detach` as a command itself.
2. Add a switch to `remove` like `--detach` or `--detach-only` which just deletes to document. 
Which one would you prefer? Which branch should the code for the feature be based on?

**Comment by humm** [comment]
I think it is important to limit the number of core commands. One solution I could see is to replace the `attach` command with a `doc ` command, with subcommands. Therefore:
- `pubs doc add $citekey $path/to/file` would add a document
- `pubs doc remove $citekey` would remove the document from the corresponding paper
- `pubs doc export $citekey` or `pubs doc copy $citekey [$path]` would copy the attached document to the given path, or the current dir if nothing is provided.

@omangin: Need your input on this. What do you think ?

For the branch, it should be developed in a separate branch, based on the current feat/new_config head. Do you want to implement this feature ?

**Comment by 73** [comment]
Sure, I'll do it. I also like the idea with the `doc` command.  
I would rather see zshcompletion done (it is my personal thing to make pubs really usable), but I really don't know where to begin.

**Comment by omangin** [comment]
I'd say that both can be implemented as aliases… but that not what we want for both.

For the command to copy the paper to the current directory I'd say that it definitely falls under alias' use case. Typically `copy = open -w 'cp -t ./'` does it.

For removing a document, that definitely needs to be in the core functions. I think the idea of the `doc` subcommand is good, thus replacing `attach`. It seems to be the path taken for the `note` subcommand too. Actually this also makes `open` quite misplaced. It should rather be `doc open`. Or `doc` as well as `note` could default to open. Then shortcuts such as `attach = doc add` or `open = doc open` can be defined in the config for those who wants.

The only thing here is that by doing so we chose a path of `pubs [obj] [cmd]` where `[obj]` would relate to the document or the notes and `[cmd]` to open, add, remove, etc. Some argument declarations and options will then be shared by the various primary commands (e.g. the `open --with`). The other path would have been to go for `pubs [cmd] [obj]`. This is closer to the `edit` command with switches for bibtex and meta files.

Although I have no certainty on the question, a nice logic would be (also the closest to what we currently have):
- top level commands for the whole paper (i.e. mandatory meta and bibtex file). Examples are `add`, `remove`, `edit`.
- subcommands for document, notes (actually this one is close to belong to plugins).
- for the `open` command It makes more sense to nest it under `doc`. Whether we need an open command for bibtex files is another story (nice for scripting but could be hacked from export. Also we may not be in favor of exposing too much the bibtex files that are essential parts of the repository. But this is kind of the philosophy of pubs too. In conclusion, let's save this one for later!)

I think we should favor the logic. Aliases can always be defined to save some keystrokes.

Sorry for the long comment.

**Comment by humm** [comment]
I agree with that. I think the `pubs doc add` syntax is the only one possible, as `pubs add doc` raises ambiguity in the parser about adding a bibtex file named `doc`. 

Also, the `open` command relies on external programs that may not be present on some platforms, or modified in the config. Using an alias for `doc copy` is not portable or robust.

@73: If you want to work on the `doc` command, it's yours. Don't worry about the 'doc copy' command if you are not interested by it.

**Comment by 73** [comment]
Hi,
I read your posts and came to think that a `doc` command would be kind of a invasive to the interface.
If we take step like it, it might be good to do all the work at once, in favour of the interface getting a little more stable.
So I looked at all commands that are there and tried to package them.
I followed the these principals:
1. Basic commands are: `init`,`add`, `edit`, `remove`, `export`, `list`, `rename` and `websearch`. They all work on the references library. The `doc` command works on the attached documents, `note` on the notes and well `tag` on tags.
2. The UI is a User Interface: So if you want to script things, there is a now basic switch `-n|--no-interaction`, where there would be a question (e.g. are you sure you want that?) is now an error shown and exit code >0. This should be nice for scripting. If one tries to `add` something that is already there, one gets an error (or a question) and shown what would be overwritten. One can do `!! -f` if one wants to force that. Removal `remove` always asks except `--force` is given. All `edit` command would bring up the editor, except `--set` is used (which forces an overwrite). `list` should bring up a pager or editor by default (so the use can search in these lists and jump around). The interface is for the user, a script might give `-p|--no-pager` to get output to `stdout`.
```
pubs ---+- [-c|--config $file] [-h|--help] [-v|--version] [-n|--no-interaction]
        |
        +--- init [-h|--help] [-p|--pubsdir $path] [-d|--docsdir $path] [-f|--force]
        |
        +--- add [$file.bib | [-D|--doi $doi] | [-I|--isdn $isbn]] [-k|--citekeys $keys] [-f|--force]
        |
        +--- edit $key [[-m|--meta] | [-s|--set $field=$value]]
        |
        +--- remove $key [$key [...]] [-f|--force]
        |
        +--- doc --+- add $file $key [[-L|--link] | [-M|--move]] [-f|--force]
        |          |
        |          +- remove $key [$key [...]] [-f|--force]
        |          |
        |          +- export $key [$path]
        |          |
        |          +- open $key [-w|--with $cmd]
        |         
        +--- note -+- add $key $note [-f|--force]
        |          |
        |          +- edit $key [-s|--set $field=$value]
        |          |
        |          +- remove $key [$key [...]] [-f|--force]
        |
        +--- tag $key [$key [...]] $tags [--rename]
        |
        +--- export [-k|--keys $keylist] [-q|--query $query] [$path]
        |
        +--- list [-k|--citekeys-only] [[-i|--ignore-case] | [-I|--force-case]] [-p|--no-pager] [-s|--sort [key|author|title|year] [-o|--order d|i]] [[$query] | [--no-docs]]
        |
        +--- rename $key $key
        |
        +--- websearch [$key | [-D|--doi $doi] | [-I|--isdn $isbn] | [-q|--query $query]]
        |
        +--- stats
``` 
The `add` command was completely overloaded with optionals stacked on optionals. This was due to the ability to add document in the same step. That's a little much isn't it? 
In my opinion that was kind of counter intuitive and leads to a steep learning curve.
I also would drop `import`, this could be done by `add` if one drops the ability to set the key but makes the key a selector.
Working on that proposal I came across the `list` command. It was the most confusing. With the `-i` and `-I` switches, which case is ignored exactly? I extended sorting on the list supporting `fields` that have to be there and increase/decrease.
I could not help myself but to make the `websearch` a little bit more useful. It can now be used to check what you would get if you would `add` a reference with the switches and to show more to a given `$key`.

This is not work done know, it is meant as a plan. I propose to rework things like that and then add completion for everything (#34). There is an issue on a `stats` command #8. What should that do?

So, what do you think?
--73

**Comment by 73** [comment]
You can completely shut me down if you guys think I went totally overboard, but I think it would be a good idea to have a "bigger picture".
--73

**Comment by humm** [comment]
There are many good ideas there. I think they should not be implemented all at once, but incrementally. Such dramatic changes would destabilise the project for a significant time (e.g., we are still dealing with instability due to changes in config), which is the opposite of what we are trying to do right now. 

Some remarks:
- The `note` command doesn't need subcommands IMO. `pubs note $citekey` open a (possibly empty) document for edit, which is both fine for consulting, adding or editing a note. And removing a note could be done the same way by removing all the content inside a note---in which case, the program could consider it the same a no-note. As such, I am a big partisan of not adding complexity there.
- The multiplication of the top-level options: "[-c|--config $file] [-h|--help] [-v|--version] [-n|--no-interaction]" raises problems, as the abbreviations preclude any use of them in subcommands. I don't think the "--version"  option should be kept. Instead, version information should appear when `pubs` is invoked alone. I would prefer "-y|--yes" to "-n|--no-interaction", shorter and more standard in my experience.
- From experience, the ability to add documents as the same time as bibtex is good. The syntax of the add function could probably be improved though. Thanks for the feedback on the learning curve, it is precious to us.
- `websearch` has always has been a bit gimmicky. I did not understand how you want to change it however.

To recap: Love the enthusiasm, and stability is paramount. Different features should be developed separately, and probably be discussed in different issue threads. For the moment, we all agree on the `doc` command, and I do think we should start with that, merge it to master, and if move forward from there.

**Comment by 73** [comment]
Yaeh, not all at once but within one version. If all agree I would make a todo list in form of multiple issues. 
* I agree on the point that `note edit` is completely redundant (I read that somehow into the comment ;). It serves the sole purpose to enable a more CRUD-like feeling for the user.  But we could do something like:  `note $key [-s|--set $note] [-f|--force]` which is a lot simpler, but also more implicit, which means more learning for the user.
* On the version part: I think it should appear in both. I assume people in general are used to `PROG --version` like I am. But I could be mistaken.
* `-y|--yes` would relate to `-f|--force` in my opinion. It forces a *yes* answer (e.i. an action to go through).  I'm completely indifferent on the name of the switch, just arguing it should be on every write action. The idea behind `-n|--no-interaction`  is that no questions gets asked, no pager or editor run. Instead exit codes and stdout are used for feedback that allows straight forward scripting (I was under the impression that scripting pubs was a thing).
* I think if the `add` command gives something back on stdout like `Successfully added $key` the user just gives `pubs doc add $file $key` is a more guided way for the user. Or hey might be unsatisfied with the `$key` so he runs `pubs rename $key $betterkey` and then `pubs doc add $file $betterkey`. Since I have to lookup the bibtex first before adding something to the database (It is not like it is extracted from the pdf), I can add some good bibtex which I looked up outside of pubs (maybe started with `websearch`) and save what I found. After that I add the document. I think it's more elegant. I get why one wants to keep it close together, but imho it is not like `pubs` thinks, `pubs` thinks bibtex first, attachments after. Maybe @omangin has an opinion on that?

**Comment by 73** [comment]
Hey,
implementing the `doc` command. And I have some questions.

First I have found something in the `repo.py` which strikes me as odd: https://github.com/pubs/pubs/blob/master/pubs/repo.py#L96-L98
Why is the Exception not raised to the cmd that invoked the command and then printed with the `ui`?

Secondly I'm not sure whether or not the removal of a document is a `RemoveEvent` in the sense of: https://github.com/pubs/pubs/blob/master/pubs/events.py One one hand I would intuitively assume not, but on the other hand I can't seem to figure out what the events are used for. They just seem to be send with nobody listening. When I extract a function `Repository.remove_doc()` from `Repository.remove_paper()` should I carry the event?

The rest seams clear so far.
Best,
--73

**Comment by humm** [comment]
On the first, as the comment said, it indeed needs to be printed by the UI. The idea there is to avoid printing a full stacktrace *or* and exception incomprehensible to most users (that don't necessarily know Python, etc.). 

Events are part of an old plugin framework, but AFAIK, nobody is using it anymore. And no, RemoveEvent should not trigger when only the document is removed. Those events should *really* have had a docstring, I'll fix it. 

Don't hesitate if you have more question.

**Comment by 73** [comment]
Thx for getting back to back to me in such a timely manner.
1. I will raise the error and print something pretty to the user.
2. I will not introduce events to any new code and leave it alone.
--73

**Comment by omangin** [comment]
Thanks for all that reading!

My comments:
- The `open` command relies on `xdg-open` or what is defined in the config so it is both portable and file type agnostic. 
- I was under the impression that copying to current directory was a very specific use case that did not justify a command. That is why I proposed an alias. Now that it has become an `export` command it seems much more useful. 
- On the `note edit` question, I use both pdf scans of handwritten notes and text files. In any case I think it is better to stay file type agnostic and let the user choose. Using `xdg-open` should do the job. It is OK too to edit a new file if none exists. I am not completely sure whether the edit command is required.
- I also use a lot the option to add the document when the paper is added. I am much in favor of keeping it. It is not a lot of code and saves a lot of typing, The bibtex file is still mandatory anyway.
- For the no interaction mode, sometime I have seen `--no-ask` too. I have no preference. On the issue of having too many letters for short commands taken by the top level ones, I would suggest not offering shortcuts for `--version` and `--no-interaction` (or `--yes` which is shorter!). The former is rarely used and the latter is only used in scripts so a few more keystrokes is not a big deal.
- Finally, I think the plugin architecture is good (although not polished) and might be used in the future. However it is maybe better to wait for the functionalities to be used before writing code for them. So let's keep it but you can skip the events for now.

Anyway, thanks for the work and enthusiasm!

**Comment by 73** [comment]
Okay. Picking up on those points I take it that:
+ If `note` should be file agnostic one needs `add`, `open` and `remove` as subcommand for `note`. I was under the impression that it would be `txt` not an arbitrary file. Then `note edit` can not open a (maybe empty) file for edit because we have no filetype.
+ Offering no short handles for `--version` and `--no-ask`is completely fine with me. `--version` is rarely used and the latter one is meant for scripting, so its fine with me.
As regarding to the other topics, I'm not that familiar with them so no opinion from my side. ;)
--73
